webroot/src/components/Users/UserRowEdit/UserRowEdit.ts

```diff
@@ -151,16 +170,42 @@ class UserRowEdit extends mixins(MixinForm) {
 
     get userOptionsState(): Array<PermissionOption> {
         const { userCompact } = this;
-        const options = userCompact?.memberStates?.map((memberState: State) => ({
+        let options = userCompact?.memberStates?.map((memberState: State) => ({
             value: (memberState.abbrev as unknown as string)?.toLowerCase() || '',
             name: memberState.name(),
         })) || [];
 
+        if (!this.isCurrentUserCompactAdmin) {
+            options = options.filter((option) =>
+                this.currentUserStatePermissions.some((statePermission) => {
+                    const stateAbbrev = statePermission.state?.abbrev || '';
+                    const isStateAdmin = statePermission.isAdmin;
+
+                    return stateAbbrev === option.value && isStateAdmin;
+                }));
+        }
+
         options.unshift({ value: '', name: this.$t('common.select') });
 
         return options;
     }
 
+    get shouldShowAddStateButton(): boolean {
+        let shouldShow = false;
+
+        if (this.isCurrentUserCompactAdmin || this.isCurrentUserStateAdminAny) {
+            const availableStatesNum = this.userOptionsState.length - 1;
+            const assignedStatesNum = Object.keys(this.formData)
+                .filter((formKey) => formKey.startsWith('state-option')).length;
+
+            if (availableStatesNum > assignedStatesNum) {
+                shouldShow = true;
```

Comment:
I think this logic is repeated, which is OK, but perhaps re: my prior request for a comment, maybe just write a comment referring to the other module in this one (or vice versa). (e.g., "see XXX.ts")